### ISSUE TITLE
Vehicle integration bugs, KAC related?

### ISSUE THREAD

**Issue by zorg2044** [Angel-125/BARIS#14]
Hi Angel, been trying out BARIS and finding it quite intriguing but I've been having several errors with the Vehicle Integration functionality

Error 1

1. Start integration
2. Minus 1 worker
3. Integration gets cancelled
4. The other workers have disappeared and need to be re-hired. Only the worker removed is available.

Error 2

Pressing the max button when assigning workers sometimes assigns more than the max.

Error 3

Integration cancelled if KAC alarm is cancelled

Error 4

1. Rush integration charges funds but integration is cancelled when rush button is pressed
2. The workers have disappeared and need to be re-hired

No. 4 was the most troublesome as the 1-3 can be avoided but 4 takes away the ability to rush.

This was on KSP 1.7.1, Windows 10, KAC v 1.3.10.0.0 and BARIS 1.9.0

I have the log file here where I loaded the game and carried out the above steps as per the video below.

https://www.dropbox.com/s/injjp97n3m4j7sr/output_log.txt?dl=0

Video
https://youtu.be/eJXh6DWyIS8

I had a thought that it might be related to Kerbal Alarm Clock so I made a copy of my install and removed  KAC and most other (but not all mods) and it seemed to be working fine. However as soon as I added KAC back in the same issues turned up.

I also tried with KAC but Kronometer removed but the issue persisted.

I hope there is a solution and this has been helpful!

**Comment by Makrond** [comment]
Integration getting cancelled when the alarm is deleted appears to be intentional behaviour, and is the root cause of all these problems (and issue #8) as the current code deletes the alarm every time it's updated.

A fix should be pretty easy, I could submit a PR if that would be helpful.

**Comment by modulozero** [comment]
The issues seem to remain: adding or removing a worker causes the entire contract to be cancelled, _and_ rushing cancels the thing. Any workers involved are instantly fired, leading to rather severe and righteous union anger. I've attached some logs:

[KSP.log](https://github.com/Angel-125/BARIS/files/4909147/KSP.log)
[Player.log](https://github.com/Angel-125/BARIS/files/4909149/Player.log)

It's really a shame, I somewhat prefer the BARIS model over KCT, it's more gamey, but in a good way, IMO.

**Comment by modulozero** [comment]
FWIW I tried Makrond's [fix](https://github.com/Makrond/BARIS/commit/fa80e3691170fb338f68466e9827a5bb460f06b3) by adding the relevant code to the most recent source, and it did fix issues 1 and 2. Issue 4 seems like an invalid max worker calculation that's a bit tricky for me to follow (I'm not particularly proficient with C#). It did lead me to adding 50 workers to a project, and then, as I removed them and then fired them:

![image](https://user-images.githubusercontent.com/144155/87257119-77129100-c498-11ea-80f7-564a9f6157c0.png)

Free workforce, yay!

**Comment by Makrond** [comment]
Yeah the short version is the code gets a bit confused about which workers go where and how many should exist sometimes. I wrote another [quick-and-dirty fix](https://github.com/Makrond/BARIS/compare/master...Makrond:worker-fixes) for that one at the same time. Fixing it "properly" would probably require a bit of restructuring and refactoring, moving all the relevant code into its own class to make it easier to reason about, and I certainly wasn't confident enough back then to try it. Edit: it's been a while so I don't remember if this also had the benefit of fixing the rush problem.
